File: android/app/src/main/java/dev/atarasy/prototype/MemberPrivateNode.kt
package dev.atarasy.prototype

import java.io.File
import java.io.FileOutputStream
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.SecureRandom
import java.util.Base64
import java.util.UUID
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.long

data class MemberPrivateNodeEnvelope(val profile: String = "atarasy.private-node-record.1", val nonce: String, val ciphertext: String)
data class MemberPrivateNodeRecord(val id: String, val revision: Long, val updatedAt: Long, val envelope: MemberPrivateNodeEnvelope)
data class MemberPrivateNodeIndex(val profile: String, val checkedAt: Long, val records: List<MemberPrivateNodeRecord>)
data class MemberPrivateNodeMove(val key: ByteArray, val source: List<MemberPrivateNodeRecord>, val target: List<MemberPrivateNodeRecord>, val clearDigests: Map<String, String>) {
    override fun equals(other: Any?) = other is MemberPrivateNodeMove && key.contentEquals(other.key) && source == other.source && target == other.target && clearDigests == other.clearDigests
    override fun hashCode() = 31 * key.contentHashCode() + source.hashCode()
}
enum class MemberPrivateNodeState { LOCKED, READY, RECOVERY_REQUIRED }

object MemberPrivateNodeCodec {
    private val json = Json { ignoreUnknownKeys = false; isLenient = false }
    fun envelope(element: JsonElement): MemberPrivateNodeEnvelope = malformed {
        val row = element.jsonObject; require(row.keys == setOf("profile", "nonce", "ciphertext"))
        val value = MemberPrivateNodeEnvelope(row.string("profile"), row.string("nonce"), row.string("ciphertext"))
        require(value.profile == "atarasy.private-node-record.1" && data(value.nonce).size == 12)
        require(data(value.ciphertext).size in 17..12_304); value
    }
    fun record(element: JsonElement): MemberPrivateNodeRecord = malformed {
        val row = element.jsonObject; require(row.keys == setOf("id", "revision", "updatedAt", "envelope"))
        val value = MemberPrivateNodeRecord(row.string("id"), row.number("revision"), row.number("updatedAt"), envelope(row.getValue("envelope")))
        require(UUID.fromString(value.id).toString() == value.id && value.revision > 0); value
    }
    fun index(bytes: ByteArray): MemberPrivateNodeIndex = malformed {
        val root = json.parseToJsonElement(bytes.toString(Charsets.UTF_8)).jsonObject
        require(root.keys == setOf("profile", "checkedAt", "records") && root.string("profile") == "atarasy.private-node-index.1")
        val rows = root.getValue("records").jsonArray.map(::record); require(rows.size <= 10_000 && rows.map { it.id }.distinct().size == rows.size && rows.map { it.id } == rows.map { it.id }.sorted())
        MemberPrivateNodeIndex("atarasy.private-node-index.1", root.number("checkedAt"), rows)
    }
    fun envelopeJson(value: MemberPrivateNodeEnvelope) = buildJsonObject {
        put("profile", JsonPrimitive(value.profile)); put("nonce", JsonPrimitive(value.nonce)); put("ciphertext", JsonPrimitive(value.ciphertext))
    }
    fun b64(bytes: ByteArray): String = Base64.getUrlEncoder().withoutPadding().encodeToString(bytes)
    fun data(value: String): ByteArray {
        if (!Regex("^[A-Za-z0-9_-]+$").matches(value) || value.length % 4 == 1) throw MemberFailure.Malformed
        return try { Base64.getUrlDecoder().decode(value).also { if (b64(it) != value) throw MemberFailure.Malformed } } catch (e: MemberFailure) { throw e } catch (_: Exception) { throw MemberFailure.Malformed }
    }
    fun scope(environment: MemberEnvironment, household: String) = Canonical.digest(JsonArray(listOf(
        JsonPrimitive("atarasy.private-node-scope.1"), JsonPrimitive(environment.name), JsonPrimitive(environment.origin), JsonPrimitive(household),
    )).toString())
    fun aad(environment: MemberEnvironment, household: String, id: String, revision: Long) = JsonArray(listOf(
        JsonPrimitive("atarasy.private-node-record.1"), JsonPrimitive(environment.name), JsonPrimitive(environment.origin), JsonPrimitive(household), JsonPrimitive(id), JsonPrimitive(revision.toString()),
    )).toString().toByteArray()
    private fun JsonObject.string(key: String) = getValue(key).jsonPrimitive.let { require(it.isString); it.content }
    private fun JsonObject.number(key: String) = getValue(key).jsonPrimitive.let { require(!it.isString); it.long }.also { require(it in 0..Canonical.MAXIMUM_INTEGER) }
    private inline fun <T> malformed(block: () -> T): T = try { block() } catch (e: MemberFailure) { throw e } catch (_: Exception) { throw MemberFailure.Malformed }
}

class MemberPrivateNodeCrypto(key: ByteArray) {
    private val secret = SecretKeySpec(key.copyOf().also { if (it.size != 32) throw MemberFailure.Storage }, "AES")
    fun seal(clear: ByteArray, environment: MemberEnvironment, household: String, id: String, revision: Long): MemberPrivateNodeEnvelope {
        if (clear.isEmpty() || clear.size > 12_288 || !canonicalId(id) || revision <= 0) throw MemberFailure.Malformed
        return try {
            val nonce = ByteArray(12).also(SecureRandom()::nextBytes); val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.ENCRYPT_MODE, secret, GCMParameterSpec(128, nonce)); cipher.updateAAD(MemberPrivateNodeCodec.aad(environment, household, id, revision))
            MemberPrivateNodeEnvelope(nonce = MemberPrivateNodeCodec.b64(nonce), ciphertext = MemberPrivateNodeCodec.b64(cipher.doFinal(clear)))
        } catch (failure: Exception) { if (failure is MemberFailure) throw failure else throw MemberFailure.Storage }
    }
    fun open(record: MemberPrivateNodeRecord, environment: MemberEnvironment, household: String): ByteArray = try {
        if (record.envelope.profile != "atarasy.private-node-record.1" || !canonicalId(record.id) || record.revision <= 0) throw MemberFailure.Malformed
        val nonce = MemberPrivateNodeCodec.data(record.envelope.nonce); val sealed = MemberPrivateNodeCodec.data(record.envelope.ciphertext)
        if (nonce.size != 12 || sealed.size !in 17..12_304) throw MemberFailure.Malformed
        val cipher = Cipher.getInstance("AES/GCM/NoPadding"); cipher.init(Cipher.DECRYPT_MODE, secret, GCMParameterSpec(128, nonce))
        cipher.updateAAD(MemberPrivateNodeCodec.aad(environment, household, record.id, record.revision)); cipher.doFinal(sealed)
    } catch (failure: Exception) { if (failure is MemberFailure) throw failure else throw MemberFailure.Storage }
    private fun canonicalId(id: String) = try { UUID.fromString(id).toString() == id } catch (_: Exception) { false }
}

interface MemberPrivateNodeKeyVault {
    fun load(scope: String): ByteArray?
    fun create(scope: String): ByteArray
    fun install(scope: String, key: ByteArray)
}

class EncryptedFilePrivateNodeKeyVault(private val directory: File, private val cipher: EnvelopeCipher) : MemberPrivateNodeKeyVault {
    private val lock = Any()
    init { if (!directory.exists() && !directory.mkdirs()) throw MemberFailure.Storage }
    override fun load(scope: String): ByteArray? = synchronized(lock) {
        validate(scope); val file = File(directory, "$scope.key"); if (!file.exists()) return@synchronized null
        try { cipher.open(file.readBytes(), aad(scope)).also { require(it.size == 32) } } catch (_: Exception) { throw MemberFailure.Storage }
    }
    override fun create(scope: String): ByteArray = synchronized(lock) {
        load(scope)?.let { return@synchronized it }; ByteArray(32).also(SecureRandom()::nextBytes).also { write(scope, it) }
    }
    override fun install(scope: String, key: ByteArray) = synchronized(lock) {
        validate(scope); if (key.size != 32 || load(scope) != null) throw MemberFailure.Storage; write(scope, key)
    }
    private fun write(scope: String, key: ByteArray) {
        val target = File(directory, "$scope.key"); val temporary = File(directory, ".$scope.${System.nanoTime()}.tmp")
        try {
            FileOutputStream(temporary).use { it.write(cipher.seal(key, aad(scope))); it.fd.sync() }
            try { Files.move(temporary.toPath(), target.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING) }
            catch (_: Exception) { Files.move(temporary.toPath(), target.toPath(), StandardCopyOption.REPLACE_EXISTING) }
        } catch (_: Exception) { temporary.delete(); throw MemberFailure.Storage }
    }
    private fun validate(scope: String) { if (!Regex("^[a-f0-9]{64}$").matches(scope)) throw MemberFailure.Storage }
    private fun aad(scope: String) = "atarasy.android-private-node-key.1\n$scope".toByteArray()
}

class MemberPrivateNodeRemote(private val sessions: MemberSessionClient) {
    suspend fun records(): MemberPrivateNodeIndex { val (reply, _) = sessions.readWithSession("/member/private-node/records"); json200(reply); return MemberPrivateNodeCodec.index(reply.body) }
    suspend fun record(id: String): MemberPrivateNodeRecord {
        canonicalId(id); val (reply, _) = sessions.readWithSession("/member/private-node/records/$id"); json200(reply)
        return parseRecord(reply.body).also { if (it.id != id) throw MemberFailure.ScopeMismatch }
    }
    suspend fun write(id: String, expectedRevision: Long, envelope: MemberPrivateNodeEnvelope): MemberPrivateNodeRecord {
        canonicalId(id); if (expectedRevision !in 0 until Canonical.MAXIMUM_INTEGER) throw MemberFailure.Malformed; MemberPrivateNodeCodec.envelope(MemberPrivateNodeCodec.envelopeJson(envelope))
        val body = buildJsonObject { put("expectedRevision", JsonPrimitive(expectedRevision)); put("envelope", MemberPrivateNodeCodec.envelopeJson(envelope)) }.toString().toByteArray()
        val (reply, _) = sessions.readWithSession("/member/private-node/records/$id", body = body); json200(reply)
        return parseRecord(reply.body).also { if (it.id != id || it.revision != expectedRevision + 1 || it.envelope != envelope) throw MemberFailure.ScopeMismatch }
    }
    private fun parseRecord(bytes: ByteArray) = try { MemberPrivateNodeCodec.record(Json.parseToJsonElement(bytes.toString(Charsets.UTF_8))) } catch (e: MemberFailure) { throw e } catch (_: Exception) { throw MemberFailure.Malformed }
    private fun canonicalId(id: String) { try { require(UUID.fromString(id).toString() == id) } catch (_: Exception) { throw MemberFailure.Malformed } }
    private fun json200(reply: MemberHttpResponse) { if (reply.status != 200) throw MemberFailure.Http(reply.status); if (reply.contentType?.substringBefore(';')?.trim()?.lowercase() != "application/json") throw MemberFailure.Malformed }
}

class MemberPrivateNode(
    private val environment: MemberEnvironment,
    private val remote: MemberPrivateNodeRemote,
    private val vault: MemberPrivateNodeKeyVault,
) {
    companion object { const val BOOTSTRAP_ID = "00000000-0000-4000-8000-000000000019"; val BOOTSTRAP = "{\"profile\":\"atarasy.private-node-bootstrap.1\"}".toByteArray() }
    private val mutex = Mutex(); private var session: MemberSessionInfo? = null; private var crypto: MemberPrivateNodeCrypto? = null
    var state = MemberPrivateNodeState.LOCKED; private set
    suspend fun open(session: MemberSessionInfo): MemberPrivateNodeState = mutex.withLock {
        lockInternal(); val index = remote.records(); val scope = MemberPrivateNodeCodec.scope(environment, session.household)
        var key = vault.load(scope)
        if (key == null && index.records.isNotEmpty()) { state = MemberPrivateNodeState.RECOVERY_REQUIRED; return@withLock state }
        if (key == null) key = vault.create(scope)
        val codec = MemberPrivateNodeCrypto(key)
        if (index.records.isEmpty()) {
            val envelope = codec.seal(BOOTSTRAP, environment, session.household, BOOTSTRAP_ID, 1)
            val record = remote.write(BOOTSTRAP_ID, 0, envelope); if (!codec.open(record, environment, session.household).contentEquals(BOOTSTRAP)) throw MemberFailure.ScopeMismatch
        } else verify(index.records, codec, session.household)
        this.session = session; crypto = codec; state = MemberPrivateNodeState.READY; state
    }
    suspend fun read(id: String): ByteArray = mutex.withLock {
        val active = session ?: throw MemberFailure.Storage; val codec = crypto ?: throw MemberFailure.Storage
        codec.open(remote.record(id), environment, active.household)
    }
    suspend fun write(id: String = UUID.randomUUID().toString(), expectedRevision: Long, clear: ByteArray): MemberPrivateNodeRecord = mutex.withLock {
        val active = session ?: throw MemberFailure.Storage; val codec = crypto ?: throw MemberFailure.Storage
        val envelope = codec.seal(clear, environment, active.household, id, expectedRevision + 1); val record = remote.write(id, expectedRevision, envelope)
        if (!codec.open(record, environment, active.household).contentEquals(clear)) throw MemberFailure.ScopeMismatch; record
    }
    suspend fun recoveryKey(expected: MemberSessionInfo): ByteArray = mutex.withLock {
        if (state != MemberPrivateNodeState.READY || session != expected) throw MemberFailure.Storage
        vault.load(MemberPrivateNodeCodec.scope(environment, expected.household)) ?: throw MemberFailure.Storage
    }
    suspend fun prepareMove(target: MemberEnvironment, expected: MemberSessionInfo): MemberPrivateNodeMove = mutex.withLock {
        val active = session ?: throw MemberFailure.Storage; val sourceCrypto = crypto ?: throw MemberFailure.Storage
        if (state != MemberPrivateNodeState.READY || active != expected || target.origin == environment.origin) throw MemberFailure.Storage
        val key = vault.load(MemberPrivateNodeCodec.scope(environment, expected.household)) ?: throw MemberFailure.Storage
        val index = remote.records(); verify(index.records, sourceCrypto, expected.household)
        val targetCrypto = MemberPrivateNodeCrypto(key); val targetRecords = mutableListOf<MemberPrivateNodeRecord>(); val digests = linkedMapOf<String, String>()
        index.records.forEach { record ->
            val clear = sourceCrypto.open(record, environment, expected.household)
            val envelope = targetCrypto.seal(clear, target, expected.household, record.id, record.revision)
            targetRecords += record.copy(envelope = envelope)
            digests[record.id] = java.security.MessageDigest.getInstance("SHA-256").digest(clear).joinToString("") { "%02x".format(it) }
        }
        MemberPrivateNodeMove(key.copyOf(), index.records, targetRecords, digests)
    }
    suspend fun installRecoveredKey(key: ByteArray, expected: MemberSessionInfo) = mutex.withLock {
        if (state != MemberPrivateNodeState.RECOVERY_REQUIRED || key.size != 32) throw MemberFailure.Storage
        val index = remote.records(); if (index.records.isEmpty()) throw MemberFailure.Storage; val codec = MemberPrivateNodeCrypto(key)
        verify(index.records, codec, expected.household); vault.install(MemberPrivateNodeCodec.scope(environment, expected.household), key)
        session = expected; crypto = codec; state = MemberPrivateNodeState.READY
    }
    suspend fun lock() = mutex.withLock { lockInternal() }
    private fun lockInternal() { session = null; crypto = null; state = MemberPrivateNodeState.LOCKED }
    private fun verify(records: List<MemberPrivateNodeRecord>, codec: MemberPrivateNodeCrypto, household: String) {
        records.forEach { record -> val clear = codec.open(record, environment, household); if (record.id == BOOTSTRAP_ID && !clear.contentEquals(BOOTSTRAP)) throw MemberFailure.ScopeMismatch }
    }
}

File: ios/AtarasyCore/Sources/AtarasyCore/MemberPrivateNode.swift
import Foundation
import CryptoKit

public struct MemberPrivateNodeEnvelope: Codable, Equatable, Sendable {
    public let profile: String
    public let nonce: String
    public let ciphertext: String
    public init(profile: String = "atarasy.private-node-record.1", nonce: String, ciphertext: String) { self.profile = profile; self.nonce = nonce; self.ciphertext = ciphertext }
}
public struct MemberPrivateNodeRecord: Codable, Equatable, Sendable {
    public let id: String
    public let revision: Int64
    public let updatedAt: Int64
    public let envelope: MemberPrivateNodeEnvelope
}
public struct MemberPrivateNodeIndex: Codable, Equatable, Sendable {
    public let profile: String
    public let checkedAt: Int64
    public let records: [MemberPrivateNodeRecord]
}

public protocol MemberPrivateNodeService: Sendable {
    func privateNodeRecords() async throws -> MemberPrivateNodeIndex
    func privateNodeRecord(id: String) async throws -> MemberPrivateNodeRecord
    func writePrivateNodeRecord(id: String, expectedRevision: Int64, envelope: MemberPrivateNodeEnvelope) async throws -> MemberPrivateNodeRecord
}
extension MemberClient: MemberPrivateNodeService {}

private func validatePrivateEnvelopeObject(_ value: Any) -> Bool {
    guard let object = value as? [String: Any], Set(object.keys) == ["profile", "nonce", "ciphertext"], object["profile"] as? String == "atarasy.private-node-record.1",
          let nonce = object["nonce"] as? String, let ciphertext = object["ciphertext"] as? String,
          (try? PrivateNodeCodec.data(nonce).count) == 12, let bytes = try? PrivateNodeCodec.data(ciphertext), bytes.count > 16, bytes.count <= 12_304 else { return false }
    return true
}
private func validatePrivateRecordObject(_ value: Any) -> Bool {
    guard let object = value as? [String: Any], Set(object.keys) == ["id", "revision", "updatedAt", "envelope"], let id = object["id"] as? String,
          UUID(uuidString: id)?.uuidString.lowercased() == id, let revision = object["revision"] as? NSNumber, let updated = object["updatedAt"] as? NSNumber,
          CFGetTypeID(revision) != CFBooleanGetTypeID(), CFGetTypeID(updated) != CFBooleanGetTypeID(), revision.int64Value > 0, updated.int64Value >= 0,
          Double(revision.int64Value) == revision.doubleValue, Double(updated.int64Value) == updated.doubleValue, validatePrivateEnvelopeObject(object["envelope"] as Any) else { return false }
    return true
}

public extension MemberClient {
    func privateNodeRecords() async throws -> MemberPrivateNodeIndex {
        let (reply, _) = try await read("/member/private-node/records")
        guard let object = try? JSONSerialization.jsonObject(with: reply.data) as? [String: Any], Set(object.keys) == ["profile", "checkedAt", "records"], object["profile"] as? String == "atarasy.private-node-index.1",
              let rows = object["records"] as? [Any], rows.count <= 10_000, rows.allSatisfy(validatePrivateRecordObject),
              let checked = object["checkedAt"] as? NSNumber, CFGetTypeID(checked) != CFBooleanGetTypeID(), checked.int64Value >= 0, Double(checked.int64Value) == checked.doubleValue else { throw MemberFailure.malformed }
        let value = try decode(MemberPrivateNodeIndex.self, reply)
        guard Set(value.records.map(\.id)).count == value.records.count, value.records.map(\.id) == value.records.map(\.id).sorted() else { throw MemberFailure.malformed }
        return value
    }
    func privateNodeRecord(id: String) async throws -> MemberPrivateNodeRecord {
        guard UUID(uuidString: id)?.uuidString.lowercased() == id else { throw MemberFailure.invalidInput }
        let (reply, _) = try await read("/member/private-node/records/" + id)
        guard let object = try? JSONSerialization.jsonObject(with: reply.data), validatePrivateRecordObject(object) else { throw MemberFailure.malformed }
        let value = try decode(MemberPrivateNodeRecord.self, reply); guard value.id == id else { throw MemberFailure.scopeMismatch }; return value
    }
    func writePrivateNodeRecord(id: String, expectedRevision: Int64, envelope: MemberPrivateNodeEnvelope) async throws -> MemberPrivateNodeRecord {
        guard UUID(uuidString: id)?.uuidString.lowercased() == id, expectedRevision >= 0, expectedRevision < 9_007_199_254_740_991,
              let envelopeObject = try? JSONSerialization.jsonObject(with: JSONEncoder().encode(envelope)), validatePrivateEnvelopeObject(envelopeObject) else { throw MemberFailure.invalidInput }
        struct Request: Encodable { let expectedRevision: Int64; let envelope: MemberPrivateNodeEnvelope }
        let (reply, _) = try await read("/member/private-node/records/" + id, body: JSONEncoder().encode(Request(expectedRevision: expectedRevision, envelope: envelope)))
        guard let object = try? JSONSerialization.jsonObject(with: reply.data), validatePrivateRecordObject(object) else { throw MemberFailure.malformed }
        let value = try decode(MemberPrivateNodeRecord.self, reply)
        guard value.id == id, value.revision == expectedRevision + 1, value.envelope == envelope else { throw MemberFailure.scopeMismatch }
        return value
    }
}

private enum PrivateNodeCodec {
    private static func canonicalArray(_ values: [String]) -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.withoutEscapingSlashes]
        return try! encoder.encode(values)
    }
    static func b64(_ data: Data) -> String { data.base64EncodedString().replacingOccurrences(of: "+", with: "-").replacingOccurrences(of: "/", with: "_").replacingOccurrences(of: "=", with: "") }
    static func data(_ value: String) throws -> Data {
        guard value.range(of: "^[A-Za-z0-9_-]+\\z", options: .regularExpression) != nil, value.count % 4 != 1 else { throw MemberFailure.malformed }
        var fixed = value.replacingOccurrences(of: "-", with: "+").replacingOccurrences(of: "_", with: "/")
        fixed += String(repeating: "=", count: (4 - fixed.count % 4) % 4)
        guard let result = Data(base64Encoded: fixed), b64(result) == value else { throw MemberFailure.malformed }
        return result
    }
    static func scope(environment: MemberEnvironment, household: String) -> String {
        SHA256.hash(data: canonicalArray(["atarasy.private-node-scope.1", environment.name, environment.origin.absoluteString, household])).map { String(format: "%02x", $0) }.joined()
    }
    static func aad(environment: MemberEnvironment, household: String, id: String, revision: Int64) -> Data {
        canonicalArray(["atarasy.private-node-record.1", environment.name, environment.origin.absoluteString, household, id, String(revision)])
    }
}

public struct MemberPrivateNodeCrypto: Sendable {
    private let key: SymmetricKey
    public init(key: Data) throws { guard key.count == 32 else { throw MemberFailure.storage }; self.key = SymmetricKey(data: key) }
    public func seal(_ clear: Data, environment: MemberEnvironment, household: String, id: String, revision: Int64) throws -> MemberPrivateNodeEnvelope {
        guard !clear.isEmpty, clear.count <= 12_288, UUID(uuidString: id)?.uuidString.lowercased() == id, revision > 0 else { throw MemberFailure.invalidInput }
        let sealed = try AES.GCM.seal(clear, using: key, authenticating: PrivateNodeCodec.aad(environment: environment, household: household, id: id, revision: revision))
        return .init(nonce: PrivateNodeCodec.b64(Data(sealed.nonce)), ciphertext: PrivateNodeCodec.b64(sealed.ciphertext + sealed.tag))
    }
    public func open(_ record: MemberPrivateNodeRecord, environment: MemberEnvironment, household: String) throws -> Data {
        guard record.envelope.profile == "atarasy.private-node-record.1", UUID(uuidString: record.id)?.uuidString.lowercased() == record.id, record.revision > 0 else { throw MemberFailure.malformed }
        let nonceData = try PrivateNodeCodec.data(record.envelope.nonce), sealed = try PrivateNodeCodec.data(record.envelope.ciphertext)
        guard nonceData.count == 12, sealed.count > 16, sealed.count <= 12_304 else { throw MemberFailure.malformed }
        let box = try AES.GCM.SealedBox(nonce: AES.GCM.Nonce(data: nonceData), ciphertext: sealed.dropLast(16), tag: sealed.suffix(16))
        return try AES.GCM.open(box, using: key, authenticating: PrivateNodeCodec.aad(environment: environment, household: household, id: record.id, revision: record.revision))
    }
}

public enum MemberPrivateNodeState: Equatable, Sendable { case locked, ready, recoveryRequired }

/// Owns decrypted node access for one foreground account. A missing local key is created only for
/// an empty node. If ciphertext already exists, a reinstall or new device must enter recovery.
public actor MemberPrivateNode {
    private static let bootstrapID = "00000000-0000-4000-8000-000000000019"
    private static let bootstrap = Data("{\"profile\":\"atarasy.private-node-bootstrap.1\"}".utf8)
    private let environment: MemberEnvironment
    private let service: any MemberPrivateNodeService
    private let vault: any MemberOperationKeyVault
    private var session: MemberSessionInfo?
    private var crypto: MemberPrivateNodeCrypto?
    public private(set) var state: MemberPrivateNodeState = .locked
    public init(environment: MemberEnvironment, service: any MemberPrivateNodeService, vault: any MemberOperationKeyVault) { self.environment = environment; self.service = service; self.vault = vault }
    private func verify(_ records: [MemberPrivateNodeRecord], codec: MemberPrivateNodeCrypto, household: String) throws {
        for record in records {
            let clear = try codec.open(record, environment: environment, household: household)
            if record.id == Self.bootstrapID, clear != Self.bootstrap { throw MemberFailure.scopeMismatch }
        }
    }
    public func open(session: MemberSessionInfo) async throws -> MemberPrivateNodeState {
        lock(); let index = try await service.privateNodeRecords(), reference = PrivateNodeCodec.scope(environment: environment, household: session.household)
        var key = try vault.key(scope: reference, create: false)
        if key == nil {
            guard index.records.isEmpty else { state = .recoveryRequired; return state }
            key = try vault.key(scope: reference, create: true)
        }
        guard let key else { throw MemberFailure.storage }
        let codec = try MemberPrivateNodeCrypto(key: key)
        if index.records.isEmpty {
            let envelope = try codec.seal(Self.bootstrap, environment: environment, household: session.household, id: Self.bootstrapID, revision: 1)
            let record = try await service.writePrivateNodeRecord(id: Self.bootstrapID, expectedRevision: 0, envelope: envelope)
            guard try codec.open(record, environment: environment, household: session.household) == Self.bootstrap else { throw MemberFailure.scopeMismatch }
        } else {
            try verify(index.records, codec: codec, household: session.household)
        }
        self.session = session; crypto = codec; state = .ready; return state
    }
    public func write(id: String = UUID().uuidString.lowercased(), expectedRevision: Int64, clear: Data) async throws -> MemberPrivateNodeRecord {
        guard state == .ready, let session, let crypto else { throw MemberFailure.storage }
        let envelope = try crypto.seal(clear, environment: environment, household: session.household, id: id, revision: expectedRevision + 1)
        let record = try await service.writePrivateNodeRecord(id: id, expectedRevision: expectedRevision, envelope: envelope)
        guard try crypto.open(record, environment: environment, household: session.household) == clear else { throw MemberFailure.scopeMismatch }
        return record
    }
    public func read(id: String) async throws -> Data {
        guard state == .ready, let session, let crypto else { throw MemberFailure.storage }
        return try crypto.open(await service.privateNodeRecord(id: id), environment: environment, household: session.household)
    }
    public func recoveryKey(session: MemberSessionInfo) throws -> Data {
        guard state == .ready, self.session?.id == session.id, self.session?.household == session.household,
              let key = try vault.key(scope: PrivateNodeCodec.scope(environment: environment, household: session.household), create: false) else { throw MemberFailure.storage }
        return key
    }
    public func installRecoveredKey(_ key: Data, session: MemberSessionInfo) async throws {
        guard state == .recoveryRequired, key.count == 32 else { throw MemberFailure.storage }
        let index = try await service.privateNodeRecords(); guard !index.records.isEmpty else { throw MemberFailure.storage }
        let codec = try MemberPrivateNodeCrypto(key: key); try verify(index.records, codec: codec, household: session.household)
        try vault.install(key: key, scope: PrivateNodeCodec.scope(environment: environment, household: session.household))
        self.session = session; crypto = codec; state = .ready
    }
    public func lock() { session = nil; crypto = nil; state = .locked }
}
